checkpoint_service.py

```diff
@@ -0,0 +1,71 @@
+import os
+import logging
+
+# List of all objectTypes that we export / import in WM
+USER_OBJECT = "users"
+GROUP_OBJECT = "groups"
+WORKSPACE_NOTEBOOK_OBJECT = "notebooks"
+WORKSPACE_DIRECTORY_OBJECT = "directories"
+WORKSPACE_NOTEBOOK_ACL_OBJECT = "acl_notebooks"
+WORKSPACE_DIRECTORY_ACL_OBJECT = "acl_directories"
+METASTORE_TABLES = "metastore"
+
+# Actions
+WM_EXPORT = "export"
+WM_IMPORT = "import"
+
+class CheckpointKeySet():
+    """Class that deals with checkpoint read and write of a particular object type"""
+
+    """
+    :param checkpoint_enabled: checkpoint enabled / disabled flag
+    :param checkpoint_file: file to read / write object keys for checkpointing
+    """
+    def __init__(self, checkpoint_enabled, checkpoint_file):
+        self._checkpoint_enabled = checkpoint_enabled
+        self._checkpoint_file = checkpoint_file
+        self._checkpoint_key_set = set()
+        self._restore_from_checkpoint_file()
+
+    # writes key to the checkpoint file
+    def write(self, key):
+        if self._checkpoint_enabled and key not in self._checkpoint_key_set:
+            with open(self._checkpoint_file, 'a+') as append_fp:
```

Comment:
You are right that this will slow down performance but I was optimizing for correctness than performance in the first iteration of the checkpoint-service as the performance impact will be minimal. But since you brought it up, let me update the PR with the change. 

Also, do you think we need to flush the buffer after every write ? System crash is a rare occurrence which will cause data loss but we are fine with some amount of data loss in case of checkpoint keys. It will just redo the actual call for those keys. Doing flush after every write will affect performance.

Reply:
I agree with that we should favor correctness over performance in this case. I'm ok to keep current implementation if you want.

On the other hand, I don't think any data loss is acceptable in WM, especially silent data loss.